src/transforms/pipelines/config.rs

```diff
@@ -42,65 +52,172 @@ impl Clone for PipelineConfig {
     }
 }
 
+#[async_trait::async_trait]
+#[typetag::serde(name = "pipeline")]
+impl TransformConfig for PipelineConfig {
+    async fn build(&self, ctx: &TransformContext) -> crate::Result<Transform> {
+        let condition = if let Some(config) = &self.filter {
+            Some(config.build(&ctx.enrichment_tables)?)
+        } else {
+            None
+        };
+
+        // Setup the interior transform chain in this pipeline.
+        //
+        // NOTE: As this pipeline is not expanding to separate transforms we
+        // lose validation that happens at the topology layer (e.g. making sure
+        // input/output types of connected transforms are compatible). This is
+        // an unfortunate consequence of having a "topology in a topology".
+        //
+        // In the future, it would be great to decouple that type of validation
+        // from the physical task boundaries. You could imagine a future where
+        // we replace transform expansion with the ability to return different
+        // types of compound transforms directly from TransformConfig::build,
+        // and those know how to play nice with a subsequent graph-based
+        // validation scheme.
+        //
+        // For now, considering that most everything is log to log with regard
+        // to pipelines we avoid adding duplicate validation here in favor of
+        // future work.
+        if self.transforms.is_empty() {
+            return Err(format!("empty pipeline: {}", self.name).into());
+        }
+        let mut transforms = Vec::with_capacity(self.transforms.len());
+        for config in &self.transforms {
+            let transform = match config.build(ctx).await? {
+                Transform::Function(transform) => Box::new(transform),
+                Transform::Synchronous(transform) => transform,
+                _ => return Err(format!("non-sync transform in pipeline: {:?}", config).into()),
+            };
+            transforms.push(transform);
+        }
+
+        let buf_in = TransformOutputsBuf::new_with_capacity(
+            vec![Output::default(DataType::all())],
+            INTERIOR_BUFFER_SIZE,
+        );
+        let buf_out = buf_in.clone();
+        Ok(Transform::Synchronous(Box::new(Pipeline {
+            condition,
+            transforms,
+            buf_in,
+            buf_out,
+        })))
+    }
+
+    fn input(&self) -> Input {
+        if let Some(transform) = self.transforms.first() {
+            transform.input()
+        } else {
+            panic!("pipeline {} does not have transforms", self.name)
+        }
```

Comment:
Repeat nit pick above about `if let Some … else`.

Is this actually reachable? If not, prefer `unreachable!` over `panic!` just to give that marker.

Reply:
It is. We do have in the soaks an example of a pipeline with no transforms, `http_datadog_filter_blackhole`. This being _valid_ actually masked a bug, [here](https://github.com/vectordotdev/vector/pull/12602/commits/4e1928a9e4cec95d2767730e24fb58ca3be86823). On that note, @binarylogic, thoughts on pipelines with no transforms? Presumably we could just elide them from the topology since they do nothing but it _feels_ very much like a bug to write a pipeline with no transforms, then silently not tell the user about it.